File: 210-course-schedule-ii/210.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        marked = {}
        in_stack = {}
        postorder = []

        DG = [[] for _ in range(numCourses)]
        for u,v in prerequisites:
            DG[v].append(u)

        def dfs(v):
            in_stack[v] = True
            marked[v] = True
            for w in DG[v]:
                if not w in marked:
                    if dfs(w):
                        return True
                else:
                    if w in in_stack and in_stack[w]:
                        return True

            in_stack[v] = False
            postorder.append(v)
            return False

        for v in range(numCourses):
            if not v in marked:
                if dfs(v):
                    return []

        return reversed(postorder)

File: 210-course-schedule-ii/210.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        vector<int> in_degree(numCourses);
        for (const vector<int>& pre : prerequisites) {
            int u = pre[1];
            int v = pre[0];
            graph[u].push_back(v);
            in_degree[v]++;
        }

        queue<int> q;
        for (int i = 0; i < numCourses; i++) {
            if (in_degree[i] == 0) {
                q.push(i);
            }
        }

        vector<int> result;
        result.reserve(numCourses);
        while (!q.empty()) {
            int curr = q.front();
            q.pop();
            result.push_back(curr);
            for (int next : graph[curr]) {
                in_degree[next]--;
                if (in_degree[next] == 0) {
                    q.push(next);
                }
            }
        }

        if (result.size() == numCourses)
            return result;
        else
            return {};
    }
};
